agent/engine/docker_task_engine.go

```diff
@@ -1535,7 +1543,33 @@ func (engine *DockerTaskEngine) stopContainer(task *apitask.Task, container *api
 		apiTimeoutStopContainer = engine.cfg.DockerStopTimeout
 	}
 
-	return engine.client.StopContainer(engine.ctx, dockerID, apiTimeoutStopContainer)
+	return engine.stopDockerContainer(dockerID, apiTimeoutStopContainer)
+}
+
+// stopDockerContainer attempts to stop the container, retrying only in case of time out errors.
+// If the maximum number of retries is reached, the container is marked as stopped. This is because docker sometimes
+// deadlocks when trying to stop a container but the actual container process is stopped.
+// for more information, see: https://github.com/moby/moby/issues/41587
+func (engine *DockerTaskEngine) stopDockerContainer(dockerID string, apiTimeoutStopContainer time.Duration) dockerapi.DockerContainerMetadata {
+	var md dockerapi.DockerContainerMetadata
+	backoff := newExponentialBackoff(stopContainerBackoffMin, stopContainerBackoffMax, stopContainerBackoffJitter, stopContainerBackoffMultiplier)
```

Comment:
This is not some http endpoint subject to rate-limiting, and we already have something like a 2.5 minute timeout for the StopContainer call to succeed, do we really need an exponential backoff here? I would say just try it again immediately or sleep for no more than 15-30 seconds.

Reply:
I would still keep the exponential backoff, so we don't issue too many calls to docker and slow it down even more.

Reply:
oh I just re-read the constants above and realize it caps at 5 seconds anyways, so I guess that's fine with me but still seems sort of pointless if the timeout is already 2.5 minutes. What's the point of backing off if the delay is orders of magnitude less than the timeout?

Reply:
what is the 2.5 minute timeout? docker stop timeout is at 30s, don't think we have a timeout for the call from engine itself.

Reply:
Does 2.5 mins timeout come from `timeout + stopContainerTimeoutBuffer ` in docker client?
[https://github.com/aws/amazon-ecs-agent/blob/9f2aa23fa01f4901d2f2eeece74db02564f3f59f/agent/dockerclient/dockerapi/docker_client.go#L721-L722](https://github.com/aws/amazon-ecs-agent/blob/9f2aa23fa01f4901d2f2eeece74db02564f3f59f/agent/dockerclient/dockerapi/docker_client.go#L721-L722)

`stopContainerTimeoutBuffer` is defined as
[https://github.com/aws/amazon-ecs-agent/blob/9f2aa23fa01f4901d2f2eeece74db02564f3f59f/agent/dockerclient/dockerapi/docker_client.go#L92-L97](https://github.com/aws/amazon-ecs-agent/blob/9f2aa23fa01f4901d2f2eeece74db02564f3f59f/agent/dockerclient/dockerapi/docker_client.go#L92-L97)
    
and `timeout` comes from `apiTimeoutStopContainer` which is passed by docker task engine
[https://github.com/aws/amazon-ecs-agent/blob/9f2aa23fa01f4901d2f2eeece74db02564f3f59f/agent/engine/docker_task_engine.go#L1533-L1538](https://github.com/aws/amazon-ecs-agent/blob/9f2aa23fa01f4901d2f2eeece74db02564f3f59f/agent/engine/docker_task_engine.go#L1533-L1538)
    
cfg .`DockerStopTimeout` has the default value `defaultDockerStopTimeout`, which is set to 30 second.

[https://github.com/aws/amazon-ecs-agent/blob/9f2aa23fa01f4901d2f2eeece74db02564f3f59f/agent/config/config_unix.go#L50-L61](https://github.com/aws/amazon-ecs-agent/blob/9f2aa23fa01f4901d2f2eeece74db02564f3f59f/agent/config/config_unix.go#L50-L61)  
  
[https://github.com/aws/amazon-ecs-agent/blob/9f2aa23fa01f4901d2f2eeece74db02564f3f59f/agent/config/config.go#L67-L68](https://github.com/aws/amazon-ecs-agent/blob/9f2aa23fa01f4901d2f2eeece74db02564f3f59f/agent/config/config.go#L67-L68)

then timeout + stopContainerTimeoutBuffer  = 30 * time.Second  +   2 * time.Minute